File: server/routes/webauthn.js
"use strict";

const crypto   = require("crypto");
const express  = require("express");
const rateLimit = require("express-rate-limit");
const router   = express.Router();

const { signToken } = require("../lib/crypto");
const requireAuth = require("../middleware/requireAuth");
const {
  readCredentials, writeCredentials,
  readPasskeyCredentials, writePasskeyCredential, deletePasskeyCredential,
  readWebAuthnChallenge, writeWebAuthnChallenge,
  readSetupState, writeSetupState,
  readRefreshTokens,
  db,
} = require("../db");

const IS_PROD          = process.env.NODE_ENV === "production";
const WEBAUTHN_RP_ID   = process.env.WEBAUTHN_RP_ID   || "localhost";
const WEBAUTHN_RP_NAME = process.env.WEBAUTHN_RP_NAME || "Account Manager";

const authLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, max: 20,
  message: { error: "Too many attempts, try again in 15 minutes" },
  standardHeaders: true, legacyHeaders: false,
  skip: () => process.env.NODE_ENV === "test",
});

let _webauthn = null;
async function getWebAuthn() {
  if (!_webauthn) _webauthn = await import("@simplewebauthn/server");
  return _webauthn;
}

function getOrigin(req) {
  return req.headers.origin || `${req.protocol}://${req.get("host")}`;
}

function saveRefreshToken(token) {
  const { readRefreshTokens, writeRefreshTokens } = require("../db");
  const tokens = readRefreshTokens();
  tokens[token] = Date.now() + 30 * 24 * 60 * 60 * 1000;
  for (const [k, exp] of Object.entries(tokens)) {
    if (exp < Date.now()) delete tokens[k];
  }
  writeRefreshTokens(tokens);
}

// ─── Registration (new user, no creds yet) ────────────────────────────────────

router.post("/webauthn/register/start", authLimiter, async (req, res) => {
  try {
    const creds = readCredentials();
    if (creds) return res.status(403).json({ error: "Already configured" });
    const { username, password } = req.body;
    if (!username || !password) return res.status(400).json({ error: "Missing fields" });
    if (password.length < 6) return res.status(400).json({ error: "Password too short" });
    const { hashPassword } = require("../lib/crypto");
    const salt = crypto.randomBytes(32).toString("hex");
    const hash = await hashPassword(password, salt);
    const { generateRegistrationOptions } = await getWebAuthn();
    const options = await generateRegistrationOptions({
      rpName: WEBAUTHN_RP_NAME,
      rpID:   WEBAUTHN_RP_ID,
      userID: new TextEncoder().encode(username.trim()),
      userName: username.trim(),
      attestationType: "none",
      authenticatorSelection: { userVerification: "preferred", residentKey: "preferred" },
    });
    writeSetupState({ username: username.trim(), hash, salt, challenge: options.challenge, createdAt: Date.now() });
    res.json(options);
  } catch (e) {
    console.error("WebAuthn register/start error:", e);
    res.status(500).json({ error: "Registration start failed" });
  }
});

router.post("/webauthn/register/finish", authLimiter, async (req, res) => {
  try {
    const creds = readCredentials();
    if (creds) return res.status(403).json({ error: "Already configured" });
    const state = readSetupState();
    if (!state || Date.now() - state.createdAt > 10 * 60 * 1000) {
      return res.status(400).json({ error: "Registration session expired" });
    }
    const { verifyRegistrationResponse } = await getWebAuthn();
    const verification = await verifyRegistrationResponse({
      response:          req.body,
      expectedChallenge: state.challenge,
      expectedOrigin:    getOrigin(req),
      expectedRPID:      WEBAUTHN_RP_ID,
    });
    if (!verification.verified) return res.status(400).json({ error: "Verification failed" });
    const { credential } = verification.registrationInfo;
    db.transaction(() => {
      writeCredentials({ username: state.username, hash: state.hash, salt: state.salt, totpSecret: "" });
      writeSetupState(null);
      writePasskeyCredential({
        credentialId: credential.id,
        publicKey:    Buffer.from(credential.publicKey).toString("base64"),
        counter:      credential.counter,
        deviceName:   req.body.deviceName || "Device 1",
        createdAt:    new Date().toISOString(),
      });
    })();
    res.json({ ok: true });
  } catch (e) {
    console.error("WebAuthn register/finish error:", e);
    res.status(500).json({ error: "Registration failed" });
  }
});

// ─── Add device (authenticated user) ─────────────────────────────────────────

router.post("/webauthn/add-device/start", requireAuth, async (req, res) => {
  try {
    const creds    = readCredentials();
    const existing = readPasskeyCredentials();
    const { generateRegistrationOptions } = await getWebAuthn();
    const options = await generateRegistrationOptions({
      rpName: WEBAUTHN_RP_NAME,
      rpID:   WEBAUTHN_RP_ID,
      userID: new TextEncoder().encode(creds.username),
      userName: creds.username,
      attestationType: "none",
      excludeCredentials: existing.map(p => ({ id: p.credentialId, type: "public-key" })),
      authenticatorSelection: { userVerification: "preferred", residentKey: "preferred" },
    });
    writeWebAuthnChallenge({ challenge: options.challenge, createdAt: Date.now() });
    res.json(options);
  } catch (e) {
    console.error("WebAuthn add-device/start error:", e);
    res.status(500).json({ error: "Failed to start" });
  }
});

router.post("/webauthn/add-device/finish", requireAuth, async (req, res) => {
  try {
    const state = readWebAuthnChallenge();
    if (!state || Date.now() - state.createdAt > 10 * 60 * 1000) {
      return res.status(400).json({ error: "Session expired" });
    }
    const { verifyRegistrationResponse } = await getWebAuthn();
    const verification = await verifyRegistrationResponse({
      response:          req.body,
      expectedChallenge: state.challenge,
      expectedOrigin:    getOrigin(req),
      expectedRPID:      WEBAUTHN_RP_ID,
    });
    if (!verification.verified) return res.status(400).json({ error: "Verification failed" });
    const { credential } = verification.registrationInfo;
    const existing = readPasskeyCredentials();
    writePasskeyCredential({
      credentialId: credential.id,
      publicKey:    Buffer.from(credential.publicKey).toString("base64"),
      counter:      credential.counter,
      deviceName:   req.body.deviceName || `Device ${existing.length + 1}`,
      createdAt:    new Date().toISOString(),
    });
    writeWebAuthnChallenge(null);
    res.json({ ok: true });
  } catch (e) {
    console.error("WebAuthn add-device/finish error:", e);
    res.status(500).json({ error: "Failed to register device" });
  }
});

// ─── Login ────────────────────────────────────────────────────────────────────

router.post("/webauthn/login/start", authLimiter, async (req, res) => {
  try {
    const passkeys = readPasskeyCredentials();
    if (passkeys.length === 0) return res.status(400).json({ error: "No passkeys registered" });
    const { generateAuthenticationOptions } = await getWebAuthn();
    const options = await generateAuthenticationOptions({
      rpID:             WEBAUTHN_RP_ID,
      allowCredentials: passkeys.map(p => ({ id: p.credentialId, type: "public-key" })),
      userVerification: "preferred",
    });
    writeWebAuthnChallenge({ challenge: options.challenge, createdAt: Date.now() });
    res.json(options);
  } catch (e) {
    console.error("WebAuthn login/start error:", e);
    res.status(500).json({ error: "Login start failed" });
  }
});

router.post("/webauthn/login/finish", authLimiter, async (req, res) => {
  try {
    const state = readWebAuthnChallenge();
    if (!state || Date.now() - state.createdAt > 5 * 60 * 1000) {
      return res.status(400).json({ error: "Authentication session expired" });
    }
    const passkeys = readPasskeyCredentials();
    const passkey  = passkeys.find(p => p.credentialId === req.body.id);
    if (!passkey) return res.status(400).json({ error: "Unknown credential" });
    const { verifyAuthenticationResponse } = await getWebAuthn();
    const verification = await verifyAuthenticationResponse({
      response:          req.body,
      expectedChallenge: state.challenge,
      expectedOrigin:    getOrigin(req),
      expectedRPID:      WEBAUTHN_RP_ID,
      credential: {
        id:        passkey.credentialId,
        publicKey: Buffer.from(passkey.publicKey, "base64"),
        counter:   passkey.counter,
      },
    });
    if (!verification.verified) return res.status(401).json({ error: "Authentication failed" });
    writePasskeyCredential({ ...passkey, counter: verification.authenticationInfo.newCounter });
    writeWebAuthnChallenge(null);
    const creds        = readCredentials();
    const accessToken  = signToken(creds.username, "account-manager", "1h");
    const refreshToken = crypto.randomBytes(48).toString("hex");
    saveRefreshToken(refreshToken);
    res.cookie("refreshToken", refreshToken, {
      httpOnly: true, secure: IS_PROD, sameSite: "strict",
      maxAge: 30 * 24 * 60 * 60 * 1000, path: "/",
    });
    res.json({ accessToken });
  } catch (e) {
    console.error("WebAuthn login/finish error:", e);
    res.status(401).json({ error: "Authentication failed" });
  }
});

router.get("/webauthn/credentials", requireAuth, (req, res) => {
  const passkeys = readPasskeyCredentials();
  res.json(passkeys.map(p => ({ credentialId: p.credentialId, deviceName: p.deviceName, createdAt: p.createdAt })));
});

router.delete("/webauthn/credentials/:id", requireAuth, (req, res) => {
  const passkeys = readPasskeyCredentials();
  if (passkeys.length <= 1) {
    return res.status(400).json({ error: "Cannot remove last passkey — register another device first" });
  }
  deletePasskeyCredential(decodeURIComponent(req.params.id));
  res.json({ ok: true });
});

module.exports = router;

File: internal/handler/webauthn.go
package handler

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"account-manager/internal/auth"
	"account-manager/internal/db"

	"github.com/go-chi/chi/v5"
	"github.com/go-webauthn/webauthn/protocol"
	gowebauthn "github.com/go-webauthn/webauthn/webauthn"
)

// waUser implements gowebauthn.User backed by our PasskeyCredential slice.
type waUser struct {
	username string
	id       []byte
	creds    []gowebauthn.Credential
}

func (u *waUser) WebAuthnID() []byte                           { return u.id }
func (u *waUser) WebAuthnName() string                         { return u.username }
func (u *waUser) WebAuthnDisplayName() string                  { return u.username }
func (u *waUser) WebAuthnCredentials() []gowebauthn.Credential { return u.creds }

func (a *App) newWebAuthn() (*gowebauthn.WebAuthn, error) {
	return gowebauthn.New(&gowebauthn.Config{
		RPID:          a.Cfg.WebAuthnRPID,
		RPDisplayName: a.Cfg.WebAuthnRPName,
		RPOrigins: []string{
			fmt.Sprintf("https://%s", a.Cfg.WebAuthnRPID),
			fmt.Sprintf("http://%s", a.Cfg.WebAuthnRPID),
		},
	})
}

// passkeysToWACreds converts stored passkeys to go-webauthn Credentials.
func passkeysToWACreds(passkeys []db.PasskeyCredential) []gowebauthn.Credential {
	var out []gowebauthn.Credential
	for _, p := range passkeys {
		raw, err := base64.StdEncoding.DecodeString(p.PublicKey)
		if err != nil {
			continue
		}
		idBytes, err := base64.RawURLEncoding.DecodeString(p.CredentialID)
		if err != nil {
			idBytes, _ = base64.StdEncoding.DecodeString(p.CredentialID)
		}
		out = append(out, gowebauthn.Credential{
			ID:        idBytes,
			PublicKey: raw,
			Authenticator: gowebauthn.Authenticator{
				SignCount: uint32(p.Counter),
			},
			Flags: gowebauthn.CredentialFlags{
				BackupEligible: p.BackupEligible,
				BackupState:    p.BackupState,
			},
		})
	}
	return out
}

func credDescriptors(creds []gowebauthn.Credential) []protocol.CredentialDescriptor {
	out := make([]protocol.CredentialDescriptor, 0, len(creds))
	for _, c := range creds {
		out = append(out, protocol.CredentialDescriptor{
			Type:         protocol.PublicKeyCredentialType,
			CredentialID: c.ID,
		})
	}
	return out
}

// POST /api/webauthn/register/start  (unauthenticated, first-run only)
func (a *App) WebAuthnRegisterStart(w http.ResponseWriter, r *http.Request) {
	count, _ := db.CountUsers(a.DB)
	if count > 0 {
		jsonErr(w, http.StatusForbidden, "Already configured")
		return
	}

	var body struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.Username == "" || body.Password == "" {
		jsonErr(w, http.StatusBadRequest, "Missing fields")
		return
	}
	if len(body.Password) < 12 {
		jsonErr(w, http.StatusBadRequest, "Password must be at least 12 characters")
		return
	}

	saltBytes := make([]byte, 32)
	if _, err := cryptoRandRead(saltBytes); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Registration start failed")
		return
	}
	salt := encodeHex(saltBytes)
	hash := auth.HashPassword(body.Password, salt)

	wa, err := a.newWebAuthn()
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Registration start failed")
		return
	}
	user := &waUser{username: body.Username, id: []byte(body.Username)}
	creation, session, err := wa.BeginRegistration(user,
		gowebauthn.WithResidentKeyRequirement(protocol.ResidentKeyRequirementPreferred),
		gowebauthn.WithAuthenticatorSelection(protocol.AuthenticatorSelection{
			UserVerification: protocol.VerificationPreferred,
		}),
	)
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Registration start failed")
		return
	}
	if err := db.WriteSetupState(a.DB, &db.SetupState{
		Username:  body.Username,
		Hash:      hash,
		Salt:      salt,
		Challenge: session.Challenge,
		CreatedAt: time.Now().UnixMilli(),
	}); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Registration start failed")
		return
	}
	jsonOK(w, creation)
}

// POST /api/webauthn/register/finish  (unauthenticated, first-run only)
func (a *App) WebAuthnRegisterFinish(w http.ResponseWriter, r *http.Request) {
	count, _ := db.CountUsers(a.DB)
	if count > 0 {
		jsonErr(w, http.StatusForbidden, "Already configured")
		return
	}
	state, _ := db.ReadSetupState(a.DB)
	if state == nil || time.Now().UnixMilli()-state.CreatedAt > 10*60*1000 {
		jsonErr(w, http.StatusBadRequest, "Registration session expired")
		return
	}

	wa, err := a.newWebAuthn()
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Registration failed")
		return
	}
	user := &waUser{username: state.Username, id: []byte(state.Username)}
	session := gowebauthn.SessionData{
		Challenge:  state.Challenge,
		UserID:     []byte(state.Username),
		CredParams: gowebauthn.CredentialParametersDefault(),
	}

	cred, err := wa.FinishRegistration(user, session, r)
	if err != nil {
		jsonErr(w, http.StatusBadRequest, "Verification failed")
		return
	}

	if err := db.CreateUser(a.DB, &db.User{
		Username:  state.Username,
		Hash:      state.Hash,
		Salt:      state.Salt,
		TotpSecret: "",
		IsAdmin:   true,
		CreatedAt: time.Now().UnixMilli(),
	}); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Registration failed")
		return
	}
	_ = db.WriteSetupState(a.DB, nil)
	if err := db.WritePasskeyCredential(a.DB, &db.PasskeyCredential{
		CredentialID:   base64.RawURLEncoding.EncodeToString(cred.ID),
		Username:       state.Username,
		PublicKey:      base64.StdEncoding.EncodeToString(cred.PublicKey),
		Counter:        int64(cred.Authenticator.SignCount),
		DeviceName:     "Device 1",
		CreatedAt:      time.Now().UTC().Format(time.RFC3339),
		BackupEligible: cred.Flags.BackupEligible,
		BackupState:    cred.Flags.BackupState,
	}); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Registration failed")
		return
	}
	jsonOK(w, map[string]bool{"ok": true})
}

// POST /api/webauthn/add-device/start  (requireAuth)
func (a *App) WebAuthnAddDeviceStart(w http.ResponseWriter, r *http.Request) {
	username := usernameFromContext(r)
	existing, _ := db.ReadPasskeyCredentialsByUser(a.DB, username)
	wa, err := a.newWebAuthn()
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Failed to start")
		return
	}
	waCreds := passkeysToWACreds(existing)
	user := &waUser{
		username: username,
		id:       []byte(username),
		creds:    waCreds,
	}
	creation, session, err := wa.BeginRegistration(user,
		gowebauthn.WithResidentKeyRequirement(protocol.ResidentKeyRequirementPreferred),
		gowebauthn.WithAuthenticatorSelection(protocol.AuthenticatorSelection{
			UserVerification: protocol.VerificationPreferred,
		}),
		gowebauthn.WithExclusions(credDescriptors(waCreds)),
	)
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Failed to start")
		return
	}
	// Use a per-session nonce so concurrent start calls don't overwrite each other.
	nonce := randomHex(16)
	purpose := "add_device:" + username + ":" + nonce
	if err := db.WriteWebAuthnSession(a.DB, purpose, session.Challenge); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Failed to start")
		return
	}
	http.SetCookie(w, &http.Cookie{
		Name:     "wa_add_nonce",
		Value:    nonce,
		HttpOnly: true,
		Secure:   a.Cfg.IsProd,
		SameSite: http.SameSiteStrictMode,
		Path:     "/",
		MaxAge:   600,
	})
	jsonOK(w, creation)
}

// POST /api/webauthn/add-device/finish  (requireAuth)
func (a *App) WebAuthnAddDeviceFinish(w http.ResponseWriter, r *http.Request) {
	username := usernameFromContext(r)
	nonceCookie, err := r.Cookie("wa_add_nonce")
	if err != nil || nonceCookie.Value == "" {
		jsonErr(w, http.StatusBadRequest, "Session expired")
		return
	}
	purpose := "add_device:" + username + ":" + nonceCookie.Value

	state, _ := db.ReadWebAuthnSession(a.DB, purpose)
	if state == nil || time.Now().UnixMilli()-state.CreatedAt > 10*60*1000 {
		jsonErr(w, http.StatusBadRequest, "Session expired")
		return
	}
	existing, _ := db.ReadPasskeyCredentialsByUser(a.DB, username)
	wa, err := a.newWebAuthn()
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Failed to register device")
		return
	}
	user := &waUser{username: username, id: []byte(username), creds: passkeysToWACreds(existing)}
	session := gowebauthn.SessionData{
		Challenge:  state.Challenge,
		UserID:     []byte(username),
		CredParams: gowebauthn.CredentialParametersDefault(),
	}

	cred, err := wa.FinishRegistration(user, session, r)
	if err != nil {
		jsonErr(w, http.StatusBadRequest, "Verification failed")
		return
	}

	deviceName := fmt.Sprintf("Device %d", len(existing)+1)
	if err := db.WritePasskeyCredential(a.DB, &db.PasskeyCredential{
		CredentialID:   base64.RawURLEncoding.EncodeToString(cred.ID),
		Username:       username,
		PublicKey:      base64.StdEncoding.EncodeToString(cred.PublicKey),
		Counter:        int64(cred.Authenticator.SignCount),
		DeviceName:     deviceName,
		CreatedAt:      time.Now().UTC().Format(time.RFC3339),
		BackupEligible: cred.Flags.BackupEligible,
		BackupState:    cred.Flags.BackupState,
	}); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Failed to register device")
		return
	}
	_ = db.DeleteWebAuthnSession(a.DB, purpose)
	http.SetCookie(w, &http.Cookie{
		Name:   "wa_add_nonce",
		Value:  "",
		MaxAge: -1,
		Path:   "/",
	})
	jsonOK(w, map[string]bool{"ok": true})
}

// POST /api/webauthn/login/start
func (a *App) WebAuthnLoginStart(w http.ResponseWriter, r *http.Request) {
	passkeys, _ := db.ReadAllPasskeyCredentials(a.DB)
	if len(passkeys) == 0 {
		jsonErr(w, http.StatusBadRequest, "No passkeys registered")
		return
	}
	wa, err := a.newWebAuthn()
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Login start failed")
		return
	}
	assertion, session, err := wa.BeginDiscoverableLogin()
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Login start failed")
		return
	}
	nonce := randomHex(16)
	if err := db.WriteWebAuthnSession(a.DB, "login:"+nonce, session.Challenge); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Login start failed")
		return
	}
	http.SetCookie(w, &http.Cookie{
		Name:     "wa_login_nonce",
		Value:    nonce,
		HttpOnly: true,
		Secure:   a.Cfg.IsProd,
		SameSite: http.SameSiteStrictMode,
		Path:     "/",
		MaxAge:   300,
	})
	jsonOK(w, assertion)
}

// POST /api/webauthn/login/finish
func (a *App) WebAuthnLoginFinish(w http.ResponseWriter, r *http.Request) {
	nonceCookie, err := r.Cookie("wa_login_nonce")
	if err != nil || nonceCookie.Value == "" {
		jsonErr(w, http.StatusBadRequest, "Authentication session expired")
		return
	}
	state, _ := db.ReadWebAuthnSession(a.DB, "login:"+nonceCookie.Value)
	if state == nil || time.Now().UnixMilli()-state.CreatedAt > 5*60*1000 {
		jsonErr(w, http.StatusBadRequest, "Authentication session expired")
		return
	}
	wa, err := a.newWebAuthn()
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Authentication failed")
		return
	}

	// FinishDiscoverableLogin resolves the real user from the credential ID,
	// fixing the userHandle mismatch that occurred with the synthetic "_all_" user approach.
	var owner *db.User
	var ownerPasskeys []db.PasskeyCredential
	handler := func(rawID, _ []byte) (gowebauthn.User, error) {
		credID := base64.RawURLEncoding.EncodeToString(rawID)
		u, err := db.GetUserByPasskeyCredentialID(a.DB, credID)
		if err != nil || u == nil {
			return nil, fmt.Errorf("unknown credential")
		}
		owner = u
		ownerPasskeys, _ = db.ReadPasskeyCredentialsByUser(a.DB, u.Username)
		return &waUser{username: u.Username, id: []byte(u.Username), creds: passkeysToWACreds(ownerPasskeys)}, nil
	}

	session := gowebauthn.SessionData{Challenge: state.Challenge}
	updatedCred, err := wa.FinishDiscoverableLogin(handler, session, r)
	if err != nil || owner == nil {
		jsonErr(w, http.StatusUnauthorized, "Authentication failed")
		return
	}

	credID := base64.RawURLEncoding.EncodeToString(updatedCred.ID)
	for _, p := range ownerPasskeys {
		if p.CredentialID == credID {
			p.Counter = int64(updatedCred.Authenticator.SignCount)
			p.BackupState = updatedCred.Flags.BackupState
			_ = db.WritePasskeyCredential(a.DB, &p)
			break
		}
	}
	_ = db.DeleteWebAuthnSession(a.DB, "login:"+nonceCookie.Value)
	http.SetCookie(w, &http.Cookie{
		Name:   "wa_login_nonce",
		Value:  "",
		MaxAge: -1,
		Path:   "/",
	})

	accessToken, csrfToken, err := a.issueSession(w, r, owner.Username)
	if err != nil {
		jsonErr(w, http.StatusInternalServerError, "Authentication failed")
		return
	}
	jsonOK(w, map[string]string{"accessToken": accessToken, "csrfToken": csrfToken})
}

// GET /api/webauthn/credentials  (requireAuth)
func (a *App) WebAuthnCredentials(w http.ResponseWriter, r *http.Request) {
	username := usernameFromContext(r)
	passkeys, _ := db.ReadPasskeyCredentialsByUser(a.DB, username)
	type view struct {
		CredentialID string `json:"credentialId"`
		DeviceName   string `json:"deviceName"`
		CreatedAt    string `json:"createdAt"`
	}
	out := make([]view, 0, len(passkeys))
	for _, p := range passkeys {
		out = append(out, view{CredentialID: p.CredentialID, DeviceName: p.DeviceName, CreatedAt: p.CreatedAt})
	}
	jsonOK(w, out)
}

// DELETE /api/webauthn/credentials/:id  (requireAuth)
func (a *App) WebAuthnDeleteCredential(w http.ResponseWriter, r *http.Request) {
	username := usernameFromContext(r)
	id := chi.URLParam(r, "id")
	if err := db.DeletePasskeyCredentialForUser(a.DB, id, username); err != nil {
		jsonErr(w, http.StatusInternalServerError, "Failed to delete credential")
		return
	}
	jsonOK(w, map[string]bool{"ok": true})
}

